README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# cpss

<!-- badges: start -->

<!-- badges: end -->

An R Package for Change-Point Detection by Sample-Splitting Methods.

## Installation

You can install the released version of cpss from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("cpss")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("ghwang-nk/cpss")
```

## Quick demos

``` r
library(cpss)
?cpss.mean
?cpss.var
?cpss.meanvar
?cpss.glm
?cpss.lm
?cpss.em
?cpss.custom
```


NEWS.md
# cpss (development version)

# cpss 0.0.2

* Resubmitted to CRAN.
  - omitted the redundant "An R Package" from the title.
  - uncommented some code lines in examples, and wrapped them in `\donttest`.

# cpss 0.0.1

* Submitted to CRAN.


R/glm.R
#' Detecting changes in GLMs
#'
#' @param formula a \code{formula} object describing the change-point model to be fitted.
#' @param family a description of the error distribution and link function to be used in the model, which can be a character string naming a family function or a family function.
#' @param data an optional data frame, list or environment containing the variables in the model.
#' @param algorithm a character string specifying the change-point searching algorithm, one of four state-of-the-art candidates "SN" (segment neighborhood), "BS" (binary segmentation), "WBS" (wild binary segmentation) and "PELT" (pruned exact linear time) algorithms.
#' @param dist_min an integer indicating the minimum distance between two successive candidate change-points, with a default value \eqn{floor(log(n))}.
#' @param ncps_max an integer indicating the maximum number of change-points searched for, with a default value \eqn{ceiling(n^0.4)}.
#' @param pelt_pen_val a numeric vector specifying the collection of candidate values of the penalty if the "PELT" algorithm is used.
#' @param pelt_K a numeric value to adjust the pruning tactic, usually is taken to be 0 if negative log-likelihood is used as a cost; more details can be found in Killick et al. (2012).
#' @param wbs_nintervals an integer indicating the number of random intervals drawn in the "WBS" algorithm and a default value 500 is used.
#' @param criterion a character string indicating which model selection criterion, "cross- validation" ("CV") or "multiple-splitting" ("MS"), is used.
#' @param times an integer indicating how many times of sample-splitting should be performed; if "CV" criterion is used, it should be set as 2.
#'
#' @return \code{cpss.glm} returns an object of an \proglang{S4} class, called "\code{cpss}", which collects data and information required for further change-point analyses and summaries. See \code{\link{cpss.custom}}.
#' @export
#'
#' @references
#' Killick, R., Fearnhead, P., and Eckley, I. A. (2012). Optimal Detection of Changepoints With a Linear Computational Cost. Journal of the American Statistical Association, 107(500):1590–1598.
#' @seealso \code{\link{cpss.lm}}
#' @examples
#' library("cpss")
#' set.seed(666)
#' n <- 200
#' size <- rpois(n, 20 - 1) + 1
#' tau <- c(75, 100, 175)
#' tau_ext <- c(0, tau, n)
#' be <- list(c(0, 0.5), c(0, -0.5), c(0.5, -0.5), c(-0.5, -0.5))
#' seg_len <- diff(c(0, tau, n))
#' x <- rnorm(n)
#' eta <- lapply(seq(1, length(tau) + 1), function(k) {
#'   be[[k]][1] + be[[k]][2] * x[(tau_ext[k] + 1):tau_ext[k + 1]]
#' })
#' eta <- do.call(c, eta)
#' p <- 1 / (1 + exp(-eta))
#' y <- rbinom(n, size = size, prob = p)
#' pelt_pen_val <- (log(n))^seq(0.5, 2, by = 0.1)
#' # res <- cpss.glm(
#' #   formula = cbind(y, size - y) ~ x, family = binomial(),
#' #   algorithm = "PELT", pelt_pen_val = pelt_pen_val,
#' #   dist_min = 5, ncps_max = 10
#' # )
#' # summary(res)
#' # 75  105  175
#' # coef(res)
#' # [1,] 0.02540872  0.08389551  0.5284425 -0.4980768
#' # [2,] 0.57222684 -0.45430385 -0.5203319 -0.4581678
#' @importFrom stats family glm model.response model.matrix
cpss.glm <- function(formula, family, data = NULL, algorithm = "BS", dist_min = floor(log(n)), ncps_max = ceiling(n^0.4), pelt_pen_val = NULL, pelt_K = 0, wbs_nintervals = 500, criterion = "CV", times = 2) {

  temp <- glm(formula, family, data, method = "model.frame")
  dataset <- cbind(model.response(temp), model.matrix(temp))
  n <- nrow(dataset)
  res <- cpss.custom(
    dataset,
    n,
    g_subdat,
    g_param_glm,
    g_cost_glm,
    algorithm,
    dist_min,
    ncps_max,
    pelt_pen_val,
    pelt_K,
    wbs_nintervals,
    criterion,
    times,
    model = "glm",
    g_smry = NULL,
    easy_cost = NULL,
    param.opt = family
  )
  res@call <- list(call = match.call())

  return(res)
}

#' Detecting changes in linear models
#'
#' @param formula a \code{formula} object describing the change-point model to be fitted.
#' @param data an optional data frame, list or environment containing the variables in the model.
#' @param algorithm a character string specifying the change-point searching algorithm, one of four state-of-the-art candidates "SN" (segment neighborhood), "BS" (binary segmentation), "WBS" (wild binary segmentation) and "PELT" (pruned exact linear time) algorithms.
#' @param dist_min an integer indicating the minimum distance between two successive candidate change-points, with a default value \eqn{floor(log(n))}.
#' @param ncps_max an integer indicating the maximum number of change-points searched for, with a default value \eqn{ceiling(n^0.4)}.
#' @param pelt_pen_val a numeric vector specifying the collection of candidate values of the penalty if the "PELT" algorithm is used.
#' @param pelt_K a numeric value to adjust the pruning tactic, usually is taken to be 0 if negative log-likelihood is used as a cost; more details can be found in Killick et al. (2012).
#' @param wbs_nintervals an integer indicating the number of random intervals drawn in the "WBS" algorithm and a default value 500 is used.
#' @param criterion a character string indicating which model selection criterion, "cross- validation" ("CV") or "multiple-splitting" ("MS"), is used.
#' @param times an integer indicating how many times of sample-splitting should be performed; if "CV" criterion is used, it should be set as 2.
#'
#' @return \code{cpss.lm} returns an object of an \proglang{S4} class, called "\code{cpss}", which collects data and information required for further change-point analyses and summaries. See \code{\link{cpss.custom}}.
#' @export
#'
#' @references
#' Killick, R., Fearnhead, P., and Eckley, I. A. (2012). Optimal Detection of Changepoints With a Linear Computational Cost. Journal of the American Statistical Association, 107(500):1590–1598.
#' @seealso \code{\link{cpss.glm}}
#' @examples
#' library("cpss")
#' set.seed(666)
#' n <- 400
#' tau <- c(80, 200, 300)
#' tau_ext <- c(0, tau, n)
#' be <- list(c(0, 1), c(1, 0.5), c(0, 1), c(-1, 0.5))
#' seg_len <- diff(c(0, tau, n))
#' x <- rnorm(n)
#' mu <- lapply(seq(1, length(tau) + 1), function(k) {
#'   be[[k]][1] + be[[k]][2] * x[(tau_ext[k] + 1):tau_ext[k + 1]]
#' })
#' mu <- do.call(c, mu)
#' sig <- unlist(lapply(seq(1, length(tau) + 1), function(k) {
#'   rep(be[[k]][2], seg_len[k])
#' }))
#' y <- rnorm(n, mu, sig)
#' res <- cpss.lm(
#'   formula = y ~ x,
#'   algorithm = "BS",
#'   dist_min = 5, ncps_max = 10
#' )
#' summary(res)
#' # 80  202  291
#' coef(res)
#' # $coef
#' # [,1]      [,2]        [,3]       [,4]
#' # [1,] -0.00188792 1.0457718 -0.03963209 -0.9444813
#' # [2,]  0.91061557 0.6291965  1.20694409  0.4410036
#' #
#' # $sigma
#' # [1] 0.8732233 0.4753216 0.9566516 0.4782329
#' @importFrom stats gaussian
cpss.lm <- function(formula, data = NULL, algorithm = "BS", dist_min = floor(log(n)), ncps_max = ceiling(n^0.4), pelt_pen_val = NULL, pelt_K = 0, wbs_nintervals = 500, criterion = "CV", times = 2) {

  family <- gaussian()
  temp <- glm(formula, family, data, method = "model.frame")
  dataset <- cbind(model.response(temp), model.matrix(temp))
  n <- nrow(dataset)
  res <- cpss.custom(
    dataset,
    n,
    g_subdat,
    g_param_glm,
    g_cost_glm,
    algorithm,
    dist_min,
    ncps_max,
    pelt_pen_val,
    pelt_K,
    wbs_nintervals,
    criterion,
    times,
    model = "lm",
    g_smry = NULL,
    easy_cost = NULL,
    param.opt = family
  )
  res@call <- list(call = match.call())

  return(res)
}


R/RcppExports.R
# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

get_cost_mean <- function(dat_smry, s, e) {
    .Call('_cpss_get_cost_mean', PACKAGE = 'cpss', dat_smry, s, e)
}

get_cost_var <- function(dat_smry, s, e) {
    .Call('_cpss_get_cost_var', PACKAGE = 'cpss', dat_smry, s, e)
}

get_cost_meanvar <- function(dat_smry, s, e) {
    .Call('_cpss_get_cost_meanvar', PACKAGE = 'cpss', dat_smry, s, e)
}

get_cost_em <- function(dat_smry, s, e) {
    .Call('_cpss_get_cost_em', PACKAGE = 'cpss', dat_smry, s, e)
}

SN <- function(dat_smry, n, L, d, model) {
    .Call('_cpss_SN', PACKAGE = 'cpss', dat_smry, n, L, d, model)
}

BS <- function(dat_smry, n, L, d, model) {
    .Call('_cpss_BS', PACKAGE = 'cpss', dat_smry, n, L, d, model)
}

WBS <- function(dat_smry, n, L, d, lr_M, model) {
    .Call('_cpss_WBS', PACKAGE = 'cpss', dat_smry, n, L, d, lr_M, model)
}

PELT <- function(dat_smry, n, pen_val, d, K, model) {
    .Call('_cpss_PELT', PACKAGE = 'cpss', dat_smry, n, pen_val, d, K, model)
}

SN_custom <- function(dat_smry, n, L, d, get_cost) {
    .Call('_cpss_SN_custom', PACKAGE = 'cpss', dat_smry, n, L, d, get_cost)
}

BS_custom <- function(dat_smry, n, L, d, get_cost) {
    .Call('_cpss_BS_custom', PACKAGE = 'cpss', dat_smry, n, L, d, get_cost)
}

WBS_custom <- function(dat_smry, n, L, d, lr_M, get_cost) {
    .Call('_cpss_WBS_custom', PACKAGE = 'cpss', dat_smry, n, L, d, lr_M, get_cost)
}

PELT_custom <- function(dat_smry, n, pen_val, d, K, get_cost) {
    .Call('_cpss_PELT_custom', PACKAGE = 'cpss', dat_smry, n, pen_val, d, K, get_cost)
}

lh_binom <- function(N, x) {
    .Call('_cpss_lh_binom', PACKAGE = 'cpss', N, x)
}

lh_multinom <- function(N, x) {
    .Call('_cpss_lh_multinom', PACKAGE = 'cpss', N, x)
}

est_cov <- function(dataset) {
    .Call('_cpss_est_cov', PACKAGE = 'cpss', dataset)
}

g_smry_mean <- function(dataset, Sigma) {
    .Call('_cpss_g_smry_mean', PACKAGE = 'cpss', dataset, Sigma)
}

g_smry_var <- function(dataset, mu) {
    .Call('_cpss_g_smry_var', PACKAGE = 'cpss', dataset, mu)
}

g_smry_meanvar <- function(dataset, param_opt) {
    .Call('_cpss_g_smry_meanvar', PACKAGE = 'cpss', dataset, param_opt)
}

g_smry_em <- function(dataset, param_opt) {
    .Call('_cpss_g_smry_em', PACKAGE = 'cpss', dataset, param_opt)
}



R/meanvar.R
#' Detecting changes in mean
#'
#' @param dataset a numeric matrix of dimension \eqn{n\times d}, where each row represents an observation and each column stands for a variable. A numeric vector could also be acceptable for univariate observations.
#' @param algorithm a character string specifying the change-point searching algorithm, one of four state-of-the-art candidates "SN" (segment neighborhood), "BS" (binary segmentation), "WBS" (wild binary segmentation) and "PELT" (pruned exact linear time) algorithms.
#' @param dist_min an integer indicating the minimum distance between two successive candidate change-points, with a default value \eqn{floor(log(n))}.
#' @param ncps_max an integer indicating the maximum number of change-points searched for, with a default value \eqn{ceiling(n^0.4)}.
#' @param pelt_pen_val a numeric vector specifying the collection of candidate values of the penalty if the "PELT" algorithm is used.
#' @param pelt_K a numeric value to adjust the pruning tactic, usually is taken to be 0 if negative log-likelihood is used as a cost; more details can be found in Killick et al. (2012).
#' @param wbs_nintervals an integer indicating the number of random intervals drawn in the "WBS" algorithm and a default value 500 is used.
#' @param criterion a character string indicating which model selection criterion, "cross- validation" ("CV") or "multiple-splitting" ("MS"), is used.
#' @param times an integer indicating how many times of sample-splitting should be performed; if "CV" criterion is used, it should be set as 2.
#' @param Sigma if a numeric matrix (or constant) is supplied, it would be taken as the value of known overall covariance (or variance). By default it is set as \code{NULL}, and the common covariance of the data is estimated based on the difference method, i.e.,
#' \deqn{\widehat{\Sigma} = \frac{1}{2(n-1)}\sum_{i=1}^{n-1} (Y_i-Y_{i+1})(Y_i-Y_{i+1})';}
#'
#' @return \code{cpss.mean} returns an object of an \proglang{S4} class, called "\code{cpss}", which collects data and information required for further change-point analyses and summaries. See \code{\link{cpss.custom}}.
#' @export
#'
#' @references
#' Killick, R., Fearnhead, P., and Eckley, I. A. (2012). Optimal Detection of Changepoints With a Linear Computational Cost. Journal of the American Statistical Association, 107(500):1590–1598.
#' @seealso \code{\link{cpss.meanvar}} \code{\link{cpss.var}}
#' @examples
#' library("cpss")
#' set.seed(666)
#' n <- 2048
#' tau <- c(205, 267, 308, 472, 512, 820, 902, 1332, 1557, 1598, 1659)
#' seg_len <- diff(c(0, tau, n))
#' mu <- rep(c(0, 14.64, -3.66, 7.32, -7.32, 10.98, -4.39, 3.29, 19.03, 7.68, 15.37, 0), seg_len)
#' ep <- 7 * rnorm(n)
#' y <- mu + ep
#' # res <- cpss.mean(y, algorithm = "SN", dist_min = 10, ncps_max = 20)
#' # summary(res)
#' # 205  267  307  471  512  820  897  1332  1557  1601  1659
#' # plot(res, type = "scatter")
#' # plot(res, type = "path")
#' # out <- update(res, dim_update = 12)
#' # out$cps_update
#' # 205  267  307  471  512  820  897 1332 1557 1601 1659 1769
#' # out$params_update
cpss.mean <- function(dataset, algorithm = "BS", dist_min = floor(log(n)), ncps_max = ceiling(n^0.4), pelt_pen_val = NULL, pelt_K = 0, wbs_nintervals = 500, criterion = "CV", times = 2, Sigma = NULL) {

  dataset <- as.matrix(dataset)
  n <- nrow(dataset)
  if (is.null(Sigma)) {
    Sigma <- est_cov(dataset)
  }
  res <- cpss.custom(
    dataset,
    n,
    g_subdat,
    g_param_mean,
    g_cost_mvnorm,
    algorithm,
    dist_min,
    ncps_max,
    pelt_pen_val,
    pelt_K,
    wbs_nintervals,
    criterion,
    times,
    model = "mean",
    g_smry = g_smry_mean,
    easy_cost = NULL,
    param.opt = Sigma
  )
  res@call <- list(call = match.call())

  return(res)
}

#' Detecting changes in (co)variance
#'
#' @param dataset a numeric matrix of dimension \eqn{n\times d}, where each row represents an observation and each column stands for a variable. A numeric vector could also be acceptable for univariate observations.
#' @param algorithm a character string specifying the change-point searching algorithm, one of four state-of-the-art candidates "SN" (segment neighborhood), "BS" (binary segmentation), "WBS" (wild binary segmentation) and "PELT" (pruned exact linear time) algorithms.
#' @param dist_min an integer indicating the minimum distance between two successive candidate change-points, with a default value \eqn{floor(log(n))}.
#' @param ncps_max an integer indicating the maximum number of change-points searched for, with a default value \eqn{ceiling(n^0.4)}.
#' @param pelt_pen_val a numeric vector specifying the collection of candidate values of the penalty if the "PELT" algorithm is used.
#' @param pelt_K a numeric value to adjust the pruning tactic, usually is taken to be 0 if negative log-likelihood is used as a cost; more details can be found in Killick et al. (2012).
#' @param wbs_nintervals an integer indicating the number of random intervals drawn in the "WBS" algorithm and a default value 500 is used.
#' @param criterion a character string indicating which model selection criterion, "cross- validation" ("CV") or "multiple-splitting" ("MS"), is used.
#' @param times an integer indicating how many times of sample-splitting should be performed; if "CV" criterion is used, it should be set as 2.
#' @param mu if a numeric vector or constant is supplied, it would be taken as the value of known overall mean. By default it is set as \code{NULL}, and the common mean of the data is estimated by the sample mean based on the entire data set.
#'
#' @return \code{cpss.var} returns an object of an \proglang{S4} class, called "\code{cpss}", which collects data and information required for further change-point analyses and summaries. See \code{\link{cpss.custom}}.
#' @export
#'
#' @references
#' Killick, R., Fearnhead, P., and Eckley, I. A. (2012). Optimal Detection of Changepoints With a Linear Computational Cost. Journal of the American Statistical Association, 107(500):1590–1598.
#' @seealso \code{\link{cpss.meanvar}} \code{\link{cpss.mean}}
#' @examples
#' library("cpss")
#' if (!requireNamespace("MASS", quietly = TRUE)) {
#'   stop("Please install the package \"MASS\".")
#' }
#' set.seed(666)
#' n <- 1000
#' tau <- c(200, 500, 750)
#' mu <- list(rep(0, 2), rep(0, 2), rep(0, 2), rep(0, 2))
#' Sigma <- list(diag(2), matrix(c(1, 0, 0, 4), 2), matrix(c(1, -0.5, -0.5, 4), 2), diag(2))
#' seg_len <- diff(c(0, tau, n))
#' y <- lapply(seq(1, length(tau) + 1), function(k) {
#'   MASS::mvrnorm(n = seg_len[k], mu = mu[[k]], Sigma = Sigma[[k]])
#' })
#' y <- do.call(rbind, y)
#' res <- cpss.var(y, algorithm = "BS", dist_min = 20)
#' cps(res)
#' # [1] 215 515 751
cpss.var <- function(dataset, algorithm = "BS", dist_min = floor(log(n)), ncps_max = ceiling(n^0.4), pelt_pen_val = NULL, pelt_K = 0, wbs_nintervals = 500, criterion = "CV", times = 2, mu = NULL) {

  dataset <- as.matrix(dataset)
  n <- nrow(dataset)
  if (is.null(mu)) {
    mu <- colMeans(dataset)
  }
  res <- cpss.custom(
    dataset,
    n,
    g_subdat,
    g_param_var,
    g_cost_mvnorm,
    algorithm,
    dist_min,
    ncps_max,
    pelt_pen_val,
    pelt_K,
    wbs_nintervals,
    criterion,
    times,
    model = "var",
    g_smry = g_smry_var,
    easy_cost = NULL,
    param.opt = mu
  )
  res@call <- list(call = match.call())

  return(res)
}

#' Detecting changes in mean and (co)variance
#'
#' @param dataset a numeric matrix of dimension \eqn{n\times d}, where each row represents an observation and each column stands for a variable. A numeric vector could also be acceptable for univariate observations.
#' @param algorithm a character string specifying the change-point searching algorithm, one of four state-of-the-art candidates "SN" (segment neighborhood), "BS" (binary segmentation), "WBS" (wild binary segmentation) and "PELT" (pruned exact linear time) algorithms.
#' @param dist_min an integer indicating the minimum distance between two successive candidate change-points, with a default value \eqn{floor(log(n))}.
#' @param ncps_max an integer indicating the maximum number of change-points searched for, with a default value \eqn{ceiling(n^0.4)}.
#' @param pelt_pen_val a numeric vector specifying the collection of candidate values of the penalty if the "PELT" algorithm is used.
#' @param pelt_K a numeric value to adjust the pruning tactic, usually is taken to be 0 if negative log-likelihood is used as a cost; more details can be found in Killick et al. (2012).
#' @param wbs_nintervals an integer indicating the number of random intervals drawn in the "WBS" algorithm and a default value 500 is used.
#' @param criterion a character string indicating which model selection criterion, "cross- validation" ("CV") or "multiple-splitting" ("MS"), is used.
#' @param times an integer indicating how many times of sample-splitting should be performed; if "CV" criterion is used, it should be set as 2.
#'
#' @return \code{cpss.meanvar} returns an object of an \proglang{S4} class, called "\code{cpss}", which collects data and information required for further change-point analyses and summaries. See \code{\link{cpss.custom}}.
#' @export
#'
#' @references
#' Killick, R., Fearnhead, P., and Eckley, I. A. (2012). Optimal Detection of Changepoints With a Linear Computational Cost. Journal of the American Statistical Association, 107(500):1590–1598.
#' @seealso \code{\link{cpss.mean}} \code{\link{cpss.var}}
#' @examples
#' library("cpss")
#' if (!requireNamespace("MASS", quietly = TRUE)) {
#'   stop("Please install the package \"MASS\".")
#' }
#' set.seed(666)
#' n <- 1000
#' tau <- c(200, 400, 600, 800)
#' mu <- list(rep(0, 2), rep(1, 2), rep(1, 2), rep(0, 2), rep(0, 2))
#' Sigma <- list(diag(2), diag(2), matrix(c(1,-1,-1, 4), 2), matrix(c(1, 0.5, 0.5, 1), 2), diag(2))
#' seg_len <- diff(c(0, tau, n))
#' y <- lapply(seq(1, length(tau) + 1), function(k) {
#'   MASS::mvrnorm(n = seg_len[k], mu = mu[[k]], Sigma = Sigma[[k]])
#' })
#' y <- do.call(rbind, y)
#' res <- cpss.meanvar(y, algorithm = "BS", dist_min = 20)
#' cps(res)
#' # [1] 211 402 598 804
#' plot(res, type = "coef")
cpss.meanvar <- function(dataset, algorithm = "BS", dist_min = floor(log(n)), ncps_max = ceiling(n^0.4), pelt_pen_val = NULL, pelt_K = 0, wbs_nintervals = 500, criterion = "CV", times = 2) {

  dataset <- as.matrix(dataset)
  n <- nrow(dataset)
  res <- cpss.custom(
    dataset,
    n,
    g_subdat,
    g_param_meanvar,
    g_cost_mvnorm,
    algorithm,
    dist_min,
    ncps_max,
    pelt_pen_val,
    pelt_K,
    wbs_nintervals,
    criterion,
    times,
    model = "meanvar",
    g_smry = g_smry_meanvar,
    easy_cost = NULL,
    param.opt = NULL
  )
  res@call <- list(call = match.call())

  return(res)
}
